File: backend/app/schemas/imports.py
from __future__ import annotations

import hashlib
import json
from datetime import datetime

from pydantic import BaseModel, Field, field_validator, model_validator

from app.domain.bundle import ArtifactStatus, OperationStatus
from app.domain.imports import ImportIntakeMode, ImportPreviewState

_TARGET_PROJECT_PATTERN = r"^[a-z0-9]+(?:[._-][a-z0-9]+)*$"


class ImportIntakeResponse(BaseModel):
    operation_id: int
    status: OperationStatus
    intake_mode: ImportIntakeMode


class MediaHandoffVerificationResponse(BaseModel):
    delivery_id: str
    signing_key_fingerprint: str
    bundle_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    bundle_size_bytes: int = Field(ge=1)
    created_at: str
    created_by: str
    verified: bool = True


class ImportDiscoveryResponse(BaseModel):
    operations: list[ImportIntakeResponse]


class ImportArtifactPreviewResponse(BaseModel):
    index: int
    artifact_type: str
    repository: str
    name: str | None = None
    reference: str | None = None
    version: str | None = None
    expected_digest: str | None = None
    target_digest: str | None = None
    payload_size: int = Field(ge=0)
    classification: ImportPreviewState
    error_code: str | None = None
    message: str | None = None


class ImportPreviewResponse(BaseModel):
    operation_id: int
    status: OperationStatus
    source_delivery_id: str
    bundle_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    bundle_size_bytes: int = Field(ge=1)
    signing_key_fingerprint: str
    verified_at: datetime

    # UI projection. Defaults keep already-persisted READY previews from older
    # v1 builds readable after an application upgrade.
    bundle_filename: str | None = None
    intake_mode: ImportIntakeMode | None = None
    source_harbor: str | None = None
    source_portal_version: str | None = None
    source_created_at: datetime | None = None
    source_created_by: str | None = None
    source_comment: str | None = None
    checksum_verified: bool = False
    signature_verified: bool = False
    schema_verified: bool = False
    overwrite_allowed: bool = False

    artifacts: list[ImportArtifactPreviewResponse]


class ImportArtifactDestinationOverride(BaseModel):
    index: int = Field(ge=0)
    target_project: str = Field(min_length=1, max_length=255, pattern=_TARGET_PROJECT_PATTERN)


class ImportDestinationPlanRequest(BaseModel):
    # Optional because intake already pins a profile for new clients. Legacy/unbound
    # operations default to "default" when first bound server-side.
    harbor_profile_id: str | None = Field(default=None, min_length=1, max_length=64)

    # Server-owned snapshot marker. Client values are overwritten by the policy-aware
    # orchestrator before planning; the field exists so persisted mapping_request is
    # cryptographically bound to the policy revision used for that plan.
    mapping_policy_revision: int = Field(default=0, ge=0)
    container_image_project: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
        pattern=_TARGET_PROJECT_PATTERN,
    )
    helm_chart_project: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
        pattern=_TARGET_PROJECT_PATTERN,
    )
    project_mappings: dict[str, str] = Field(default_factory=dict)
    artifact_overrides: list[ImportArtifactDestinationOverride] = Field(default_factory=list)

    @field_validator("project_mappings")
    @classmethod
    def validate_project_mappings(cls, value: dict[str, str]) -> dict[str, str]:
        import re

        normalized: dict[str, str] = {}
        pattern = re.compile(_TARGET_PROJECT_PATTERN)
        for source, target in value.items():
            if not pattern.fullmatch(source) or not pattern.fullmatch(target):
                raise ValueError("project mappings must use normalized Harbor project names")
            normalized[source] = target
        return normalized

    @field_validator("artifact_overrides")
    @classmethod
    def reject_duplicate_override_indices(
        cls,
        value: list[ImportArtifactDestinationOverride],
    ) -> list[ImportArtifactDestinationOverride]:
        indices = [item.index for item in value]
        if len(indices) != len(set(indices)):
            raise ValueError("artifact override indices must be unique")
        return value


class ImportDestinationArtifactPlanResponse(BaseModel):
    index: int = Field(ge=0)
    artifact_type: str
    source_repository: str
    source_project: str
    name: str | None = None
    reference: str | None = None
    version: str | None = None
    expected_digest: str | None = None
    payload_size: int = Field(ge=0)
    target_project: str | None = None
    target_repository: str | None = None
    final_reference: str | None = None
    project_exists: bool = False
    write_allowed: bool = False
    target_digest: str | None = None
    classification: ImportPreviewState
    error_code: str | None = None
    message: str | None = None


def destination_plan_id(
    bundle_sha256: str,
    artifacts: list[ImportDestinationArtifactPlanResponse],
) -> str:
    """Stable identity of bundle + resolved mapping, excluding observed TARGET state."""
    identity = {
        "bundle_sha256": bundle_sha256,
        "artifacts": [
            {
                "index": item.index,
                "artifact_type": item.artifact_type,
                "source_repository": item.source_repository,
                "source_project": item.source_project,
                "name": item.name,
                "reference": item.reference,
                "version": item.version,
                "expected_digest": item.expected_digest,
                "target_project": item.target_project,
                "target_repository": item.target_repository,
                "final_reference": item.final_reference,
            }
            for item in sorted(artifacts, key=lambda item: item.index)
        ],
    }
    encoded = json.dumps(identity, sort_keys=True, separators=(",", ":")).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


class ImportDestinationPlanResponse(BaseModel):
    operation_id: int
    source_delivery_id: str
    actor_username: str
    bundle_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    plan_id: str = Field(pattern=r"^[a-f0-9]{64}$")
    plan_hash: str = Field(pattern=r"^[a-f0-9]{64}$")
    mapping_policy_revision: int = Field(default=0, ge=0)
    created_at: datetime
    valid: bool
    artifacts: list[ImportDestinationArtifactPlanResponse]

    @model_validator(mode="after")
    def normalize_derived_fields(self) -> ImportDestinationPlanResponse:
        # plan_id is deliberately stable across TARGET state drift and policy revisions
        # when the resolved destinations remain identical. plan_hash binds the revision.
        self.plan_id = destination_plan_id(self.bundle_sha256, self.artifacts)
        self.valid = all(
            item.project_exists
            and item.write_allowed
            and item.classification
            in {ImportPreviewState.NEW, ImportPreviewState.SAME, ImportPreviewState.CONFLICT}
            for item in self.artifacts
        )
        return self


class ImportExecuteRequest(BaseModel):
    overwrite_conflicts: bool = False
    destination_plan_id: str | None = Field(default=None, pattern=r"^[a-f0-9]{64}$")


class ImportStartResponse(BaseModel):
    operation_id: int
    status: OperationStatus


class ImportReceiptArtifactResponse(BaseModel):
    index: int
    artifact_type: str
    repository: str
    name: str | None = None
    reference: str | None = None
    version: str | None = None
    expected_digest: str | None = None
    target_digest: str | None = None
    target_repository: str | None = None
    final_reference: str | None = None
    status: ArtifactStatus
    error_code: str | None = None
    error_message: str | None = None


class ImportReceiptResponse(BaseModel):
    operation_id: int
    source_delivery_id: str
    bundle_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    actor_username: str
    started_at: datetime
    finished_at: datetime
    overwrite_conflicts: bool
    destination_plan_id: str | None = Field(default=None, pattern=r"^[a-f0-9]{64}$")
    destination_plan_hash: str | None = Field(default=None, pattern=r"^[a-f0-9]{64}$")
    retry_of_operation_id: int | None = Field(default=None, gt=0)
    failure_policy: str | None = None
    result: str
    artifacts: list[ImportReceiptArtifactResponse]

File: frontend/src/api/imports.ts
import type { AxiosProgressEvent } from 'axios'

import { apiClient } from '@/api/client'
import {
  apiErrorInfo,
  cancelOperation,
  getOperation,
  type ApiErrorInfo,
  type ArtifactStatus,
  type Operation,
  type OperationStatus,
} from '@/api/exports'

export { apiErrorInfo, cancelOperation, getOperation }
export type { ApiErrorInfo, ArtifactStatus, Operation, OperationStatus }

export type BrowserPhysicalHandoffFiles = {
  bundle: File
  sidecar: File
  handoff: File
}

function fileToBase64(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onerror = () => reject(reader.error ?? new Error('Не удалось прочитать metadata file'))
    reader.onload = () => {
      const result = typeof reader.result === 'string' ? reader.result : ''
      const separator = result.indexOf(',')
      if (separator < 0) {
        reject(new Error('Не удалось закодировать metadata file'))
        return
      }
      resolve(result.slice(separator + 1))
    }
    reader.readAsDataURL(file)
  })
}

export type ImportIntakeMode = 'upload' | 'incoming'
export type ImportPreviewState = 'NEW' | 'SAME' | 'CONFLICT' | 'UNKNOWN' | 'ERROR'

export type MediaHandoffVerification = {
  delivery_id: string
  signing_key_fingerprint: string
  bundle_sha256: string
  bundle_size_bytes: number
  created_at: string
  created_by: string
  verified: boolean
}

export type ImportIntake = {
  operation_id: number
  status: OperationStatus
  intake_mode: ImportIntakeMode
}

export type ImportDiscovery = {
  operations: ImportIntake[]
}

export type ImportArtifactPreview = {
  index: number
  artifact_type: string
  repository: string
  name: string | null
  reference: string | null
  version: string | null
  expected_digest: string | null
  target_digest: string | null
  payload_size: number
  classification: ImportPreviewState
  error_code: string | null
  message: string | null
}

export type ImportPreview = {
  operation_id: number
  status: OperationStatus
  source_delivery_id: string
  bundle_sha256: string
  bundle_size_bytes: number
  signing_key_fingerprint: string
  verified_at: string
  bundle_filename: string | null
  intake_mode: ImportIntakeMode | null
  source_harbor: string | null
  source_portal_version: string | null
  source_created_at: string | null
  source_created_by: string | null
  source_comment: string | null
  checksum_verified: boolean
  signature_verified: boolean
  schema_verified: boolean
  overwrite_allowed: boolean
  artifacts: ImportArtifactPreview[]
}

export type ImportArtifactDestinationOverride = {
  index: number
  target_project: string
}

export type ImportDestinationPlanRequest = {
  harbor_profile_id?: string | null
  container_image_project: string | null
  helm_chart_project: string | null
  project_mappings: Record<string, string>
  artifact_overrides: ImportArtifactDestinationOverride[]
}

export type ImportDestinationArtifactPlan = {
  index: number
  artifact_type: string
  source_repository: string
  source_project: string
  name: string | null
  reference: string | null
  version: string | null
  expected_digest: string | null
  payload_size: number
  target_project: string | null
  target_repository: string | null
  final_reference: string | null
  project_exists: boolean
  write_allowed: boolean
  target_digest: string | null
  classification: ImportPreviewState
  error_code: string | null
  message: string | null
}

export type ImportDestinationPlan = {
  operation_id: number
  source_delivery_id: string
  actor_username: string
  bundle_sha256: string
  plan_id: string
  plan_hash: string
  mapping_policy_revision: number
  created_at: string
  valid: boolean
  artifacts: ImportDestinationArtifactPlan[]
}

export type ImportRetry = {
  operation_id: number
  retry_of_operation_id: number
  status: OperationStatus
  failure_policy: 'continue-on-error'
  destination_plan: ImportDestinationPlan
}

export type ImportReceiptArtifact = {
  index: number
  artifact_type: string
  repository: string
  name: string | null
  reference: string | null
  version: string | null
  expected_digest: string | null
  target_digest: string | null
  target_repository: string | null
  final_reference: string | null
  status: ArtifactStatus
  error_code: string | null
  error_message: string | null
}

export type ImportReceipt = {
  operation_id: number
  source_delivery_id: string
  bundle_sha256: string
  actor_username: string
  started_at: string
  finished_at: string
  overwrite_conflicts: boolean
  skip_conflicts: boolean
  destination_plan_id: string | null
  destination_plan_hash?: string | null
  retry_of_operation_id?: number | null
  failure_policy?: string | null
  result: string
  artifacts: ImportReceiptArtifact[]
}

export async function verifyPhysicalHandoff(
  file: File,
): Promise<MediaHandoffVerification> {
  const response = await apiClient.post<MediaHandoffVerification>(
    '/imports/handoff/verify',
    file,
    {
      headers: {
        'Content-Type': 'application/json',
      },
    },
  )
  return response.data
}

export async function uploadImportBundle(
  file: File,
  onProgress?: (loaded: number, total: number | null) => void,
  physicalHandoff?: Pick<BrowserPhysicalHandoffFiles, 'sidecar' | 'handoff'>,
  profileId?: string,
): Promise<ImportIntake> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/gzip',
  }
  if (physicalHandoff) {
    headers['X-HTP-Bundle-Filename'] = file.name
    headers['X-HTP-Sidecar-Base64'] = await fileToBase64(physicalHandoff.sidecar)
    headers['X-HTP-Handoff-Base64'] = await fileToBase64(physicalHandoff.handoff)
  }
  const response = await apiClient.post<ImportIntake>('/imports/upload', file, {
    headers,
    params: profileId ? { profile_id: profileId } : undefined,
    timeout: 0,
    onUploadProgress: (event: AxiosProgressEvent) => {
      onProgress?.(event.loaded, event.total ?? null)
    },
  })
  return response.data
}

export async function discoverImportBundles(profileId?: string): Promise<ImportDiscovery> {
  const response = await apiClient.post<ImportDiscovery>('/imports/discover', undefined, {
    params: profileId ? { profile_id: profileId } : undefined,
  })
  return response.data
}

export async function getImportPreview(operationId: number): Promise<ImportPreview> {
  const response = await apiClient.get<ImportPreview>(`/imports/${operationId}/preview`)
  return response.data
}

export async function buildImportDestinationPlan(
  operationId: number,
  mapping: ImportDestinationPlanRequest,
): Promise<ImportDestinationPlan> {
  const response = await apiClient.put<ImportDestinationPlan>(
    `/imports/${operationId}/destination-plan`,
    mapping,
  )
  return response.data
}

export async function prepareImportRetry(
  operationId: number,
  destinationPlanId: string,
): Promise<ImportRetry> {
  const response = await apiClient.post<ImportRetry>(`/imports/${operationId}/retry`, {
    destination_plan_id: destinationPlanId,
  })
  return response.data
}

export async function executeImport(
  operationId: number,
  overwriteConflicts: boolean,
  destinationPlanId: string,
  skipConflicts = false,
): Promise<{ operation_id: number; status: OperationStatus }> {
  const response = await apiClient.post<{ operation_id: number; status: OperationStatus }>(
    `/imports/${operationId}/execute`,
    {
      overwrite_conflicts: overwriteConflicts,
      skip_conflicts: skipConflicts,
      destination_plan_id: destinationPlanId,
    },
  )
  return response.data
}

export async function getImportReceipt(operationId: number): Promise<ImportReceipt> {
  const response = await apiClient.get<ImportReceipt>(`/imports/${operationId}/receipt`)
  return response.data
}
